### ISSUE TITLE
CSS variable substitution

### ISSUE THREAD

**Issue by chasm** [parcel-bundler/parcel-css#69]
I'm looking for a CSS parser that will replace CSS variables with their values (accounting for the cascade) for performance reasons.

I have CSS that looks like this throughout my stylesheets:

```css
input.email-field-input {
  border: var(--email-input-border,
    var(--input-border,
      var(--border,
        1px solid red
      )
    )
  );
}
```

If in a theme file I define only one of these (for example):

```css
:root {
  --input-border: 2px dotted blue;
}
```

Then I want to be able to run the parser and end up with:

```css
input.email-field-input {
  border: 2px dotted blue;
}
```

Variables gone &ndash; correct values substituted.

But then if I define the same variable further down, say:

```css
input {
  --input-border: 3px dashed yellow;
}
```

Then I would expect the compiled CSS to change accordingly:

```css
input.email-field-input {
  border: 3px dashed yellow;
}
```

And, of course, if I defined none of the variables at all, I'd expect the final default:

```css
input.email-field-input {
  border: 1px solid red;
}
```

Does Parcel-CSS do this? If not, is it or can it be under consideration as an enhancement?

I am still working out the kinks in this idea for theming a site, but however it works out, CSS variables are a performance hit I'd like to avoid. And unless they are dynamically generated, there is no point in not doing substitution in a static CSS file.

Thanks!

**Comment by devongovett** [comment]
It's generally not possible to do this without knowing the DOM structure. Variables are inherited by child elements, so without knowing what those elements are, we cannot do this properly. For example:

```html
<body>
  <div class="a">
    <div class="b"></div>
  </div>
  <div class="c"></div>
</body>
```

```css
body {
  --color: green;
}

.a {
  --color: red;
}

.b, .c {
  color: var(--color);
}
```

In this case, b should be red and c should be green. With the CSS alone it's not possible to determine this.

The only case where it could be possible would be variables defined in the `:root` selector.

**Comment by chasm** [comment]
Yeah, after I posted that I was thinking the same thing. And only possible for variables in the :root selector if they were guaranteed not to be redefined lower in the cascade.

So it could potentially be done in the browser (has to be, actually), but even there not if there are dynamic updates.

In my situation, I actually do have the DOM available, but it sounds like a lot of work.

Thanks!

**Comment by devongovett** [comment]
Going to close this then.

**Comment by bwindels** [comment]
Just wanted to point out a use case for this:
I need to support IE11, but still want to use css variables, especially to help with theming support.
I would love to switch over to `parcel-css` one day, but right now I use `postcss` with `postcss-css-variables` to do this.
This works, but you indeed need to restrict yourself to use css variables in selectors that `postcss-css-variables` can deal with, more specifically:

```css
.parent {
  --some-var: blue;
}

.parent .child {
  color: var(--some-var);
}
```
 It's better than not being able to use css variables at all though.

**Comment by ScottAwesome** [comment]
@devongovett  I know this is closed, however I wanted to ask is there any reason we wouldn't want do this via configuration? `postcss-css-variables` does manage to do this quite well, and works (arguably, maybe thats why you don't have to do this, however being in Rust it would be far far far more performant). 

I understand the issue with not knowing things ahead of time. One nice alternative to this perhaps would be hoisting these variables to `:root` at least, so that something like `css-vars-ponyfill` can handle the variable references?


Having substitution of custom properties off by default is totally reasonable given the arguments here. However I think it would be a great addition to the community to support at least variable hoisting if not variable substitution entirely. Would you be opposed to that?

Basically the idea being this

```css
/* Before */
body {
  --color: green;
}


.a {
  color: var(--color);
}

.b {
  --color: red;
}

.c, .d {
  color: var(--color);
}
```


```css
:root {
  --color1x232: green;
 --color2x232: red;
}

/* You could even remove empty selectors here */
body {}

.a {
  color: var(--color1x232);
}

.b {
  --color: red;
}

.c, .d {
  color: var(--color2x232);
}
```

**Comment by devongovett** [comment]
What's the usecase for this? All browsers support CSS variables natively now. IE 11 is no longer supported by Microsoft.

**Comment by chris-morgan** [comment]
What people mostly *want* in these cases is compile-time variables. Root-level custom properties with the explicitly-declared understanding that the bundle is complete and nothing else wants to touch the variables, so that they can be inlined everywhere, is just the most practical standard syntax and method for this.

The goals then are:

1. Minification and performance: the values of most of these sorts of variables are distinctly shorter than their names (e.g. `var(--body-padding-left)` → `1rem`, `var(--color-primary)` → `#123456`). Inlining allows further minification too (e.g. `calc(var(--body-padding-left) + var(--body-padding-right))` → `2rem`). Inlining will fairly obviously improve style computation performance too by reducing indirection.

2. Combining it with other forms of syntax lowering: e.g. `color-mix(in srgb, var(--x) 50%, transparent)` or `rgb(from var(--x), r g b / calc(alpha * 0.5)` to halve the opacity of a predefined colour `--x`. (See also #112.)

I’d be quite open to some approach other than custom properties, some approach where values are explicitly to be evaluated and inlined at compile-time before anything else happens. Preferably something that can be defined inside a stylesheet, but for many purposes something in the config would be acceptable too. #193 (design tokens) shows one approach that’s somewhat more capable.

But given just a philosophy of standards-compliant CSS and making it work in insufficiently-capable environments, it would be *really* useful to be able to declare “custom properties defined on `:root` are not used outside of this stylesheet and should therefore be inlined”. This would be enough for most people.

**Comment by ocavue** [comment]
I've created a lightningcss plugin to inline CSS variable during compile-time:

https://github.com/ocavue/lightningcss-plugin-css-variables
